python/tank/bootstrap/manager.py

```diff
@@ -221,49 +247,160 @@ def set_progress_callback(self, callback):
         """
         self._progress_cb = callback
 
-    def bootstrap_engine(self, engine_name, entity=None):
+    def bootstrap_engine(self,
+                         engine_name,
+                         entity=None,
+                         progress_callback=None,
+                         completed_callback=None,
+                         failed_callback=None):
         """
         Create an sgtk instance for the given project or site,
         then launch into the given engine.
 
+        When property ``async_bootstrap`` is false, the whole engine bootstrap logic
+        will be executed synchronously in the main application thread. When true,
+        an :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+
         If entity is None, the method will bootstrap into the site
         config. This method will attempt to resolve the config according
         to business logic set in the associated resolver class and based
         on this launch a configuration. This may involve downloading new
         apps from the toolkit app store and installing files on disk.
 
+        3 callback functions can be provided.
+
+        A callback function that reports back on the bootstrap progress
+        with the following signature::
+
+            progress_callback(step_number, message, current_index, maximum_index)
```

Comment:
In order to implement a progress toolbar in the plug-in client, a progress value that increases for the whole duration of the bootstrap process was needed. Though, at Josh's suggestion, I have replaced `step_number `by `progress_value`, a float number ranging from 0.0 to 1.0 representing the (step based, not time based) percentage of work completed.

In method `_report_progress`, callbacks with both the classic and the new signatures are supported by examining the number of callback arguments with `self._progress_cb.func_code.co_argcount`.